modules/network_synchronizer/networked_controller.cpp

```diff
@@ -0,0 +1,1336 @@
+}
+
+bool NetworkedController::is_nonet_controller() const {
+	ERR_FAIL_COND_V(get_tree() == nullptr, false);
+	if (controller_type != CONTROLLER_TYPE_NULL)
+		return controller_type == CONTROLLER_TYPE_NONET;
+	return get_tree()->get_network_peer().is_null();
+}
+
+void NetworkedController::set_inputs_buffer(const BitArray &p_new_buffer) {
+	inputs_buffer.get_buffer_mut().get_bytes_mut() = p_new_buffer.get_bytes();
+}
+
+void NetworkedController::set_scene_rewinder(SceneRewinder *p_rewinder) {
+	scene_rewinder = p_rewinder;
+}
+
+SceneRewinder *NetworkedController::get_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+bool NetworkedController::has_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+void NetworkedController::_rpc_server_send_inputs(Vector<uint8_t> p_data) {
+	ERR_FAIL_COND(get_tree()->is_network_server() == false);
+
+	const Vector<int> &peers = get_active_doll_peers();
+	for (int i = 0; i < peers.size(); i += 1) {
+		// This is an active doll, Let's send the data.
+		const int peer_id = peers[i];
+		rpc_unreliable_id(peer_id, "_rpc_doll_send_inputs", p_data);
+	}
+
+	controller->receive_inputs(p_data);
+}
+
+void NetworkedController::_rpc_send_tick_additional_speed(int p_speed) {
+	ERR_FAIL_COND(is_player_controller() == false);
+
+	real_t additional_speed = (static_cast<real_t>(p_speed) / 100.0) * MAX_ADDITIONAL_TICK_SPEED;
+	additional_speed = CLAMP(additional_speed, -MAX_ADDITIONAL_TICK_SPEED, MAX_ADDITIONAL_TICK_SPEED);
+
+	PlayerController *player_controller = static_cast<PlayerController *>(controller);
+	player_controller->tick_additional_speed = additional_speed;
+}
+
+void NetworkedController::_rpc_doll_send_inputs(Vector<uint8_t> p_data) {
+	ERR_FAIL_COND_MSG(get_tree()->is_network_server() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+	ERR_FAIL_COND_MSG(is_network_master() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+
+	controller->receive_inputs(p_data);
+}
+
+void NetworkedController::_rpc_doll_notify_connection_status(bool p_open) {
+	ERR_FAIL_COND_MSG(get_tree()->is_network_server() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+	ERR_FAIL_COND_MSG(is_network_master() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+
+	if (p_open) {
+		static_cast<DollController *>(controller)->open_flow();
+	} else {
+		static_cast<DollController *>(controller)->close_flow();
+	}
+}
+
+void NetworkedController::process(real_t p_delta) {
+	if (controller) {
+		// This is called by the `sceneRewinder` that it's not aware about the
+		// controller state; so check that the controller is not null.
+		controller->physics_process(p_delta);
```

Comment:
Is it really okay to call this in `process`? Since the function is called `physics_process` I would expect it to only run during the physics integration phase. I'm not sure that it is okay to call this from `process` which runs every frame.

Reply:
I need to rename this to avoid confusions, but the function `process` is called within the `physics_process`